### ISSUE TITLE
OSError [Error 29] llegal seek

### ISSUE THREAD

**Issue by wpf710** [benoitc/gunicorn#1038]
Hi

   I have a python application on hand ,based on falcon(https://github.com/falconry/falcon),  and actually , I used the application as a proxy to access another one.

  in the application,  it just using requests.request to access another url and feed back current response
  when I run the application with 

gunicorn "abc.wsgi:make_api()" --bind="127.0.0.1:8000" --timeout=600 -w 1

  and run the 'curl -i http://127.0.0.1:8000/', it will failed with 

  2015-05-24 03:45:10 [129721] [ERROR] Error handling request
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/workers/sync.py", line 134, in handle_request
    resp.write_file(respiter)
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/http/wsgi.py", line 341, in write_file
    fd_offset = os.lseek(fileno, 0, os.SEEK_CUR)
OSError: [Errno 29] Illegal seek

  any if I write a verify simple app and run with gunicorn , I can't reproduce it 

  so what do you think can cause such error?

**Comment by benoitc** [comment]
Does your app send a file over the socket? If yes, is the user launching gunicorn is able to read it.

**Comment by wpf710** [comment]
I don't think it send a file. and  with debugging , I saw that my app had already get the expected content as below.

actually, the following URL and content are my expected content through my app as a proxy, it return Json. 

norman@ubuntu:~$ curl -i http://192.168.94.129:5000/
HTTP/1.1 300 Multiple Choices
Date: Sun, 24 May 2015 12:42:43 GMT
Server: Apache/2.4.7 (Ubuntu)
Vary: X-Auth-Token
Content-Length: 597
Content-Type: application/json

{"versions": {"values": [{"status": "stable", "updated": "2015-03-30T00:00:00Z", "media-types": [{"base": "application/json", "type": "application/vnd.openstack.identity-v3+json"}], "id": "v3.4", "links": [{"href": "http://192.168.94.129:5000/v3/", "rel": "self"}]}, {"status": "stable", "updated": "2014-04-17T00:00:00Z", "media-types": [{"base": "application/json", "type": "application/vnd.openstack.identity-v2.0+json"}], "id": "v2.0", "links": [{"href": "http://192.168.94.129:5000/v2.0/", "rel": "self"}, {"href": "http://docs.openstack.org/", "type": "text/html", "rel": "describedby"}]}]}}

and in my app, the main code as below:

class OpenstackStream(object):
    def __init__(self, stream, size=None):
        self.stream = stream
        self.size = size

    def __len__(self):
        return self.size

    def read(self, size=None):
        return self.stream.read(size=size)

    def __iter__(self):
        return self.stream.__iter__()

    def __next__(self):
        return self.stream.__next__()
    next = __next__


class OpenStackResponder(object):
.....................
    def _standard_responder(self, req, resp, **_):
        data = None
        if (req.method == 'POST' or req.method == 'PUT'):
            if req.content_length:
                data = OpenstackStream(req.stream, size=req.content_length)

        ...............
        os_resp = requests.request(req.method,
                                   endpoint,
                                   data=data,
                                   headers=req.headers,
                                   stream=True)

        resp.status = os_resp.status_code
        content_type = os_resp.headers.pop(
            'Content-Type', 'application/json').split(';', 1)[0]

        # Hack for some reason
        if content_type == 'text/html':
            content_type = 'text/plain; charset=UTF-8'
        resp.content_type = content_type
        resp.stream_len = os_resp.headers.pop('Content-Length', 0)
        resp.set_headers(os_resp.headers)
        resp.stream = os_resp.raw

    on_get = _standard_responder

**Comment by benoitc** [comment]
The error happen because your application is passing the `wsgi.file_wrapper`setting to the environ. The value given should be a [file-like object](https://www.python.org/dev/peps/pep-0333/#optional-platform-specific-file-handling).

Normally Gunicorn should detect if the file is a real file  or not:
https://github.com/benoitc/gunicorn/blob/master/gunicorn/util.py#L511-L521

Do you have any example application I could check? Which library does it use to handle WSGI?

**Comment by wpf710** [comment]
I did not have an example application for you yet.

The library is falcon (https://github.com/falconry/falcon) if I understand right.

after grep the code,  it seem to be related with the code I paste above :
  ```
  resp.stream=os_resp.raw
```
 
  since in the falcon response.py  
```
        stream: Either a file-like object with a `read()` method that takes
            an optional size argument and returns a block of bytes, or an
            iterable object, representing response content, and yielding
            blocks as byte strings. Falcon will use *wsgi.file_wrapper*, if
            provided by the WSGI server, in order to efficiently serve
            file-like objects.
```

and  I sat that it (falcon) also just refer to env[wsgi.file_wrapper]

so I am wondering that who set the env value?

**Comment by benoitc** [comment]
Gunicorn provides the file wrapper but it expects a file-like object. It may be something missing in the object provided by falcon or a bug in Gunicorn. I am currently investing it.

**Comment by wpf710** [comment]
thank a lot

**Comment by benoitc** [comment]
Turn out that this issue is caused by request. The raw object provided [implements](https://github.com/kennethreitz/requests/blob/master/requests/packages/urllib3/response.py#L68) the `tell`and `fileno`s functions which let gunicorn thinks that itùs a real file object.

I am not sure where we should fix it right now. Also I wonder why this object is providing the `tell` function since you can' go back in the past. cc @kennethreitz . One temporary solution is to disable the sendfile option. For the fix I am still thinking of it. Maybe we could try an except on the first seek?

**Comment by tilgovi** [comment]
We seek on the `fileno()` response. Not sure how that could cause any problem since requests will throw an error when we try to call that if it can't return us an actual fd.

**Comment by tilgovi** [comment]
If that wasn't clear, I'm not sure how we could get to the line where this errors unless we have a valid `fd`.

**Comment by benoitc** [comment]
@tilgovi Not sure I follow but the raw object is quite like a wrapper over the socket. You can't seek over a socket which create the error above imo.

**Comment by benoitc** [comment]
and this error happen because the sendfile API is used if we expect a file there.

**Comment by tilgovi** [comment]
Right. A socket has a file descriptor but isn't seekable. Of course.

I was just trying to point out that the lines of code we have in util are:

```python
def is_fileobject(obj):
    if not hasattr(obj, "tell") or not hasattr(obj, "fileno"):
        return False

    # check BytesIO case and maybe others
    try:
        obj.fileno()
    except (IOError, io.UnsupportedOperation):
        return False

    return True
```

Requests is fine. There's no reason why we can't `tell` where we are in a stream that isn't seekable. It throws an exception if there's no `fileno`. The probably is definitely that we assume a fileno means we can seek. Maybe instead of a try/catch around the `lseek` in `http` we should be doing it in `is_fileobject`.

Or, it also occurs to me to ask why we call `lseek` at all. The first one seems wholly superfluous because we can get the same information from `tell` (and we do, on the very next line). Then we `lseek` at the end. Is there some spec that says we have to leave the file as we found it? In all cases where we use this function we finish the request right afterward. Why rewind the file?

**Comment by benoitc** [comment]
I need to check for linux but te thing is that `sendfile(1)` can only be used to stream a file to a socket. This why we should test if a file is given or not at first hand. 

Anyway I  agree there is nothing preventing a file-object to expose the `tell`and `fileno` methods/properties. So it let us to try/catch on the first `seek` I guess except if there is an api that would allows us to test if we're using a file. At least that way we make sure we have file behind so we can use the `sendfile`api. 

About removing the seeks, that sounds like a good idea, just at the moment I don't remember the reason for them, i need to look at my notes if I've still have them. I propose in between to do the following:

1. Patch the `is_fileobject` function to make the first seek
2. Open a ticket to remove the seeks that revert to the last position if they are useless

Thoughts?

**Comment by tilgovi** [comment]
Depends on the schedule. If we can remove the seeks and it won't take too much time to figure that out, that seems better to me. If we patch it with a try/catch and remove those later it's not particularly helpful for @wpf710 unless we also make a release. If we don't, @wpf710 can do that patch themselves while we figure out whether these seeks are necessary.

**Comment by tilgovi** [comment]
So, if it matters that much, mark it as a blocker and on a milestone for the next release. If we hit a deadline and want to release we can try/catch. If, before that, someone investigates and can explain why we do or don't need them, then we'll know.

**Comment by benoitc** [comment]
oh i think the try catch is here to stay. Don't get me wrong :) this is
only about the second part. i would do that fix now.

**Comment by tilgovi** [comment]
Don't get me wrong. Defensive programming is great. But in this case we are talking about a try/catch around a line that might not need to exist. So if that line goes away, the try/catch certainly will ;)

**Comment by tilgovi** [comment]
So, yeah, throw in the try/catch. I'll take a look and see whether it's necessary at all.

**Comment by tilgovi** [comment]
Opened #1040 and assigned to me to look into it.

**Comment by benoitc** [comment]
@tilgovi I saw your point :) Thanks for taking care about it. I will provide a patch later in the day that implement the try/catch solution

**Comment by wpf710** [comment]
@benoitc  Hi what's the status for it? 
Thanks

**Comment by tilgovi** [comment]
I've opened #1094 to address this issue. The root problem is that it is not possible to use sendfile with a socket as the source. In addressing this issue, I've simplified the sendfile logic quite a bit and made sure that a response is still generated even if the file descriptor reports a size of zero with `os.fstat()`. I think this will be more robust. I've tested it lightly, locally, but if anyone can try this out in a real-world scenario that would be most helpful.
